### DISCUSSION TITLE
How should panics be handled?

### DISCUSSION THREAD

**Discussion by kirillsalykin** [maxcountryman/underway#52] [Q&A]
Currently if task fails with a panic - it is stuck (eg it being retried and retried).

I would expect any panic should just mark task as `failed`.

On related note, I think there should be a way to mark step as failed,   like `To::failure()`

**Comment by maxcountryman** [comment]
I agree: as it stands, panics are simply not handled. I suspect we may want to try to catch panics so that we can set the state to failed--I haven't had the time to look into this.

> On related note, I think there should be a way to mark step as failed, like To::failure()

How would you like this to pair with `task::Error::Fatal`? Currently, a task can return `Err(Error::Fatal("oh no"))` and the worker will mark it as failed accordingly. This could be encapsulated with `To` as well.
